src/server/pfs/server/driver.go

```diff
@@ -1407,64 +1407,199 @@ func (d *driver) deleteCommit(ctx context.Context, commit *pfs.Commit) error {
 	if err := d.checkIsAuthorized(ctx, commit.Repo, auth.Scope_WRITER); err != nil {
 		return err
 	}
+	// inspect 'commit' to retrieve the commit ID (we'll read it again in a txn)
 	commitInfo, err := d.inspectCommit(ctx, commit, false)
 	if err != nil {
 		return err
 	}
+	commitID := commitInfo.Commit.ID
 
-	if commitInfo.Finished != nil {
-		return fmt.Errorf("cannot delete finished commit")
-	}
+	// Main txn: Delete all downstream commits, and update subvenance of upstream commits
+	deleted := make(map[string]*pfs.CommitInfo, len(commitInfo.Subvenance)) // deleted commits
+	deleteScratch := false                                                  // only delete scratch if txn succeeds
+	if _, err := col.NewSTM(ctx, d.etcdClient, func(stm col.STM) error {
+		repos := d.repos.ReadWrite(stm)
 
-	// Delete the scratch space for this commit
-	prefix, err := d.scratchCommitPrefix(ctx, commit)
-	if err != nil {
-		return err
-	}
-	_, err = d.etcdClient.Delete(ctx, prefix, etcd.WithPrefix())
-	if err != nil {
-		return err
-	}
+		// re-read CommitInfo inside txn
+		commits := d.commits(commit.Repo.Name).ReadWrite(stm)
+		commitInfo = &pfs.CommitInfo{}
+		if err := commits.Get(commitID, commitInfo); err != nil {
+			return err
+		}
+		deleteScratch = commitInfo.Finished == nil
 
-	// If this commit is the head of a branch, make the commit's parent
-	// the head instead.
-	branchInfos, err := d.listBranch(ctx, commit.Repo)
-	if err != nil {
-		return err
-	}
+		// Helper for deleting commits
+		deleteCommit := func(repo col.ReadWriteCollection, commit *pfs.Commit) error {
+			// Store commitInfo in 'deleted' and remove commit from etcd
+			commitInfo := &pfs.CommitInfo{}
+			if err := repo.Get(commit.ID, commitInfo); err != nil {
+				return err
+			}
+			deleted[commit.ID] = commitInfo
+			if err := repo.Delete(commit.ID); err != nil {
+				return err
+			}
 
-	for _, branchInfo := range branchInfos {
-		if branchInfo.Head != nil && branchInfo.Head.ID == commitInfo.Commit.ID {
-			if err := d.createBranch(ctx, branchInfo.Branch, commitInfo.ParentCommit, nil); err != nil {
+			// Update repo size
+			// TODO this is basically wrong. Other commits could share data with the
+			// commit being removed, in which case we're subtracting too much. Doing
+			// the right thing is hard, though. Hopefully nobody ends up with a
+			// negative repo size and cares too much
+			repoInfo := &pfs.RepoInfo{}
+			if err := repos.Get(commit.Repo.Name, repoInfo); err != nil {
+				return err
+			}
+			repoInfo.SizeBytes -= commitInfo.SizeBytes
```

Comment:
How do we compute commit sizes? If we're computing it as the "new" data that's not present in the parent commit then this is pretty close to correct. I don't think we are but maybe we could.

Reply:
I think it's the size of the whole tree (`driver.go:784`) but yeah I agree that we could. Should I make that change too?

Reply:
Yeah if it seems easy I'd go for it. 1 clarification though, it seems like there's 2 options here, we could start storing the new size for commit sizes and use that value directly or we could store what we've been storing and compute the new size when a commit gets deleted. I'm inclined toward the first one since it leaves the meaning of commit sizes the same but that might just be inertia on my part, I could also see an argument that displaying the new size for commits is better.